h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -872,30 +626,75 @@ protected static SparseMatrixDimensions calculateCSRMatrixDimensions(Chunk[] chu
         return new SparseMatrixDimensions(nonZeroElementsCount, ++rowIndicesCount);
     }
 
-    protected static SparseMatrixDimensions calculateCSRMatrixDimensions(Frame f, int[] chunks, Vec.Reader[] vecs, Vec.Reader w, DataInfo di) {
-        long nonZeroElementsCount = 0;
-        long rowIndicesCount = 0;
+    static SparseMatrixDimensions calculateCSRMatrixDimensions(Frame f, int[] chunkIds, Vec w, DataInfo di) {
+        CalculateCSRMatrixDimensionsMrFun fun = new CalculateCSRMatrixDimensionsMrFun(f, di, w, chunkIds);
+        H2O.submitTask(new LocalMR(fun, chunkIds.length)).join();
+        
+        return new SparseMatrixDimensions(ArrayUtils.sum(fun._nonZeroElementsCounts), ArrayUtils.sum(fun._rowIndicesCounts) + 1);
```

Comment:
The reason to do the sum here and not in the reduce phase of the `CalculateCSRMatrixDimensionsMrFun` is speed, right ? `ArrayUtils.sum` is just one method invocation, but `reduce` is many -> creating many contexts and destroying them quickly. For high number of chunks, this could get faster.

Maybe I'm getting into premature optimization, but for higher number of chunks, the number of additions done in a single thread might be as big as `Integer.MAX_VALUE - k`, where k is a small constant (typically `8`) - basically the MAX_ARRAY_SIZE allowed by the platform. Still, this operation is trivial `O(n)`, with parallelization we could achieve `O(n/t + t) + k`, where `t` is the number of threads and `k` is a simplification constant representing overhead used for thread management. This divide & conquer approach could theoretically bring us better benchmark results on large data sets.

Reply:
I will explain the (2) first because it will make (1) easier to justify.

ad 2) This particular function is called during model training. Every model training starts with making sure the chunk layout of the frame is not hurtful to per-chunk parallelization, this basically means there are enough chunks for available number of CPU cores and the chunks are not too small. Ideal layout will have something like 4x #cores chunks per Vec. This means number of local chunks per Vec is low in the optimal case (hundreds) but there is enough work for each core to do something for most of the time of the execution of the "local MRTask". The amount of work is also comparable - the ideal distribution will have the same number of rows per chunk. This makes per-chunk parallelization work well in H2O because all CPU cores all working for most of the time. If the per-chunk map() is optimal then the whole utilization of the CPU is also optimal.

The length of array `fun._nonZeroElementsCounts` is the same as number of local chunks. Because this is executed in model training we know we can rely on almost optimal chunk layout and we are dealing with just hundreds of numbers to calculate the sum. Calculating the sum is thus trivial in terms of execution time compared to what is done in the LocalMR task.

ad 1) I could use reduce() but I chose not to because of 2 reasons:

a) it is easier to see in debugger what is going on and because of (1) we don't have any performance benefit doing it either way.

b) the next step of this PR is actually to *populate the matrix itself in parallel*. This could be done if keep the pointers (the 2 arrays) and then use them as indices to the emulated infinite array (the 2D structure). This is a similar approach to how we build the BigDenseMatrix, only difference is that we need to calculate the pointers in a different way. That is why I am keeping the pointers unaggregated - I can use them later for to accomplish this goal. And it doesn't hurt performance because the sizes of the metadata arrays are too small to matter.

Reply:
Great, makes sense. Thank you !